zip-proportional-output-fee-mechanism-pofm.rst

```diff
@@ -0,0 +1,244 @@
+transactions with 2 outputs.
+
+
+Requirements for gathering consensus
+------------------------------------
+
+Wallet developers SHOULD update the fees to the proposed formula under Specification section
+by Madars Virza and Kris Nuttycombe [#madars-1]_ 
+
+Where #inputs and #outputs also take into account transparent inputs and outputs. 
+Otherwise, the fee structure (if not otherwise changed) will preferentially encourage 
+usage of the transparent part of the protocol.
+
+The change to the conventional transaction fees must be undertaken soon
+as the Zcash network has been under heavy load with high-output transactions while 
+regular shielded transactions with 2 outputs are affected when relying on current 
+light wallet infrastructure.
+
+The following parties need to be part of the consensus:
+
+* The technical aspects of a changing conventional fee based on outputs 
+need to be evaluated.
+* A guarantee from mining groups is required to include the updated POFM 
+transactions in the next block.
+* Wallet developers need to update the software to use the new fee.
+* Zcash documentation and community outreach must be undertaken to
+make the change known.
+
+
+Requirements for adoption
+-------------------------
+
+The change to the conventional transaction fees should be undertaken soon
+as it gets difficult to gain consensus with the growth in the network
+of wallets, exchanges, miners, and third parties involved.
+
+The following parties need to be part of the consensus:
+
+* Support from mining groups is required to include the updated conventional
+  fee transactions in the next block.
+* Wallet developers need to provide a commitment to update the software to use
+  the new fee.
+* Zcash documentation and community outreach must be undertaken to make the
+  change known.
+* Changes to librustzcash crates to:
+  * expose minimum recommended fee for a given number of recipients
+  * implement default fee equation and corresponding unit tests
+  * integration of such mechanism to Transaction Builder APIs
+* adopt `librustzcash` changes to mobile SDK repos:
+  * ZcashLightClientKit
+  * zcash-light-client-ffi
+  * zcash-android-wallet-sdk
+
+
+Specification
+=============
+
+Wallets implementing this specification will use a conventional fee in the form of 
+base_fee and marginal_fee.
+
+The proposal would end up selecting one of:
+* base_fee = 1000, marginal_fee = 250 in @nuttycom's proposal.
+* base_fee = 1000, marginal_fee = 1000 in @madars' proposal.
+* base_fee = 10000, marginal_fee = 2500 in @daira's proposal.
+* base_fee = 1000, marginal_fee = 1000 for Shielded, Shielding and De-shielding
+transactions, and base_fee = 10000, marginal_fee = 10000 for Transparent transactions 
+per @nighthawk24's proposal.
+
+And calculated per the following forumla:
+min_fee = base_fee + marginal_fee * max(0, inputs + outputs - grace_window_size)
+
+where grace_window_size = 4
```

Comment:
Edited to clarify the specification, and incorporate sections of ZIP 313 that need to be preserved / updated (as this ZIP would be replacing ZIP 313):

```suggestion
This specification defines three parameters that are used to calculate the
conventional fee:

=================== ============================
Parameter           Units
=================== ============================
`base_fee`          zatoshis
`marginal_fee`      zatoshis per input or output
`grace_window_size` inputs or outputs
=================== ============================

Wallets implementing this specification SHOULD use a conventional fee
calculated in zatoshis per the following formula::

    base_fee + marginal_fee * max(0, inputs + outputs - grace_window_size)

The parameters are set to the following values:
* `base_fee = TODO`;
* `marginal_fee = TODO`;
* `grace_window_size = 4`.

It is not a consensus requirement that fees follow this formula; however,
wallets SHOULD use this fee to reduce information leakage in transactions
unless overridden by the user.

Transaction relaying
--------------------

zcashd, zebrad, and potentially other node implementations, implement
fee-based restrictions on relaying of mempool transactions. Nodes that
normally relay transactions are expected to do so for transactions that pay
at least the conventional fee, unless there are other reasons not to do so
for robustness or denial-of-service mitigation.

Mempool size limiting
---------------------

zcashd and zebrad limit the size of the mempool as described in [#zip-0401]_.
This specifies a *low\_fee\_penalty* that is added to the "eviction weight"
if the transaction pays a fee less than the conventional transaction fee.
This threshold is modified to use the new conventional fee formula.

Block production
----------------

Miners, mining pools, and other block producers, select transactions for
inclusion in blocks using a variety of criteria. Where the criteria
previously used the conventional transaction fee defined in ZIP 313 to
decide on transaction inclusion, it is expected to instead use the formula
specified in this ZIP.

Open Issues
-----------

> TODO: Remove this section once a decision is made.

Possible alternatives for the parameters:
* base_fee = 1000, marginal_fee = 250 in @nuttycom's proposal.
* base_fee = 1000, marginal_fee = 1000 in @madars' proposal.
* base_fee = 10000, marginal_fee = 2500 in @daira's proposal.
* base_fee = 1000, marginal_fee = 1000 for Shielded, Shielding and De-shielding
  transactions, and base_fee = 10000, marginal_fee = 10000 for Transparent transactions 
  per @nighthawk24's proposal.
```